companion.py

```diff
@@ -545,176 +696,357 @@ def auth_callback(self) -> None:
+            # is also lagging.
+            if (last_starport := station_data.get('lastStarport')) is None:
+                logger.error("No lastStarport in data!")
+                return station_data
+
+            if ((last_starport_name := last_starport.get('name')) is None
+                    or last_starport_name == ''):
+                # This could well be valid if you've been out exploring for a long
+                # time.
+                logger.warning("No lastStarport name!")
+                return station_data
+
+            # WORKAROUND: n/a | 06-08-2021: Issue 1198 and https://issues.frontierstore.net/issue-detail/40706
+            # -- strip "+" chars off star port names returned by the CAPI
+            last_starport_name = last_starport["name"] = last_starport_name.rstrip(" +")
+
+            services = last_starport.get('services', {})
+            if not isinstance(services, dict):
+                # Odyssey Alpha Phase 3 4.0.0.20 has been observed having
+                # this be an empty list when you've jumped to another system
+                # and not yet docked.  As opposed to no services key at all
+                # or an empty dict.
+                logger.error(f'services is "{type(services)}", not dict !')
+                # TODO: Change this to be dependent on its own CL arg
+                if __debug__:
+                    self.dump_capi_data(station_data)
+
+                # Set an empty dict so as to not have to retest below.
+                services = {}
+
+            last_starport_id = int(last_starport.get('id'))
+
+            if services.get('commodities'):
+                market_data = capi_single_query(self.FRONTIER_CAPI_PATH_MARKET, timeout=timeout)
+                if last_starport_id != int(market_data['id']):
+                    logger.warning(f"{last_starport_id!r} != {int(market_data['id'])!r}")
+                    raise ServerLagging()
 
-        self.retrying = False
-        if 'timestamp' not in data:
-            data['timestamp'] = time.strftime('%Y-%m-%dT%H:%M:%SZ', parsedate(r.headers['Date']))  # type: ignore
+                else:
+                    market_data['name'] = last_starport_name
+                    station_data['lastStarport'].update(market_data)
 
-        # Update Odyssey Suit data
-        if endpoint == URL_QUERY:
-            self.suit_update(data)
+            if services.get('outfitting') or services.get('shipyard'):
+                shipyard_data = capi_single_query(self.FRONTIER_CAPI_PATH_SHIPYARD, timeout=timeout)
+                if last_starport_id != int(shipyard_data['id']):
+                    logger.warning(f"{last_starport_id!r} != {int(shipyard_data['id'])!r}")
+                    raise ServerLagging()
 
-        return data
+                else:
+                    shipyard_data['name'] = last_starport_name
+                    station_data['lastStarport'].update(shipyard_data)
+            # WORKAROUND END
```

Comment:
Not entirely sure this should be here?

Reply:
It's the same coverage of the WORKAROUND as in `develop`.  Ask past A_D why they had the end of it be that far down ;)

Reply:
Oh. well that guy obviously made a mistake. I thiiink? Not sure. probably leave it in that case there was probably a reason